go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -154,76 +285,268 @@ func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add bo
+	if err != nil {
+		klog.V(5).Infof("No endpoint found for service %s in namespace %s", service.Name, service.Namespace)
+		return
+	}
+
+	if _, exists := npw.getServiceInfo(name); !exists {
+		klog.V(5).Infof("Rules already deleted for service %s in namespace %s", service.Name, service.Namespace)
+	} else {
+		npw.DeleteEndpoints(ep)
+	}
 }
 
 func (npw *nodePortWatcher) SyncServices(services []interface{}) {
+	keepIPTRules := []iptRule{}
 	for _, serviceInterface := range services {
+		name := ktypes.NamespacedName{Namespace: serviceInterface.(*kapi.Service).Namespace, Name: serviceInterface.(*kapi.Service).Name}
+
 		service, ok := serviceInterface.(*kapi.Service)
 		if !ok {
 			klog.Errorf("Spurious object in syncServices: %v",
 				serviceInterface)
 			continue
 		}
-		npw.updateServiceFlowCache(service, true)
-	}
 
+		ep, err := npw.watchFactory.GetEndpoint(service.Namespace, service.Name)
+		if err != nil {
+			klog.V(5).Infof("No endpoint found for service %s in namespace %s during sync", service.Name, service.Namespace)
+			continue
+		}
+
+		hasHostNet := hasHostNetworkEndpoints(ep, &npw.nodeIPManager.addresses)
+		npw.setServiceInfo(name, service, hasHostNet)
+		// Delete OF rules for service if they exist
+		npw.updateServiceFlowCache(service, false, hasHostNet)
+		npw.updateServiceFlowCache(service, true, hasHostNet)
+		// Add correct iptables rules
+		keepIPTRules = append(keepIPTRules, getGatewayIPTRules(service, nil, hasHostNet)...)
+	}
+	// sync OF rules once
 	npw.ofm.requestFlowSync()
-	syncSharedGatewayIptRules(services)
+	// sync IPtables rules once
+	for _, chain := range []string{iptableNodePortChain, iptableExternalIPChain} {
+		recreateIPTRules("nat", chain, keepIPTRules)
+	}
+}
+
+// Add new Local Traffic flows if endpoints are local
+func (npw *nodePortWatcher) AddEndpoints(ep *kapi.Endpoints) {
+	name := ktypes.NamespacedName{Namespace: ep.Namespace, Name: ep.Name}
+
+	svc, err := npw.watchFactory.GetService(ep.Namespace, ep.Name)
+	if err != nil {
+		// This is not necessarily an error. For e.g when there are endpoints
+		// without a corresponding service.
+		klog.V(5).Infof("No service found for endpoint %s in namespace %s", ep.Name, ep.Namespace)
+		return
+	}
+	klog.V(5).Infof("Matching service %s found for ep: %s, with cluster IP: %s", svc.Name, ep.Name, svc.Spec.ClusterIP)
+
+	if !util.ServiceTypeHasClusterIP(svc) || !util.IsClusterIPSet(svc) {
+		return
+	}
+
+	etpHostRules := hasHostNetworkEndpoints(ep, &npw.nodeIPManager.addresses)
+	npw.addServiceRules(svc, etpHostRules)
+	npw.setServiceInfo(name, svc, etpHostRules)
+}
+
+func (npw *nodePortWatcher) DeleteEndpoints(ep *kapi.Endpoints) {
+	name := ktypes.NamespacedName{Namespace: ep.Namespace, Name: ep.Name}
+	// Always deletes rules for existing configuration
+	if svcConfig, exists := npw.getAndDeleteServiceInfo(name); exists {
+		npw.delServiceRules(svcConfig.service, svcConfig.etpHostRules)
```

Comment:
if the service still exists, and just the endpoints are gone. You still need some rules